e2e/testsuite/testconfig.go

```diff
@@ -207,6 +216,35 @@ func (tc TestConfig) validateGenesisDebugConfig() error {
 	return err
 }
 
+// validateUpgradeConfig ensures the upgrade configuration is valid.
+func (tc TestConfig) validateUpgradeConfig() error {
+	if strings.TrimSpace(tc.UpgradePlanName) == "" {
+		return nil
+	}
+
+	// the upgrade plan name specified must match one of the upgrade plans in the upgrade configs.
+	foundPlan := false
+	for _, upgrade := range tc.UpgradeConfigs {
+		if strings.TrimSpace(upgrade.Tag) == "" {
+			return fmt.Errorf("upgrade config missing tag: %+v", upgrade)
+		}
+
+		if strings.TrimSpace(upgrade.PlanName) == "" {
+			return fmt.Errorf("upgrade config missing plan name: %+v", upgrade)
+		}
+
+		if upgrade.PlanName == tc.UpgradePlanName {
+			foundPlan = true
+		}
```

Comment:
[nit] if we go for either 1 or 2 in the above comment, we can `return nil` here.

Or even:

if we go with 1 above: use `slices.IndexFunc()` to get the index of the right upgrade plan, and then do the validation and return

if we go with 2 above: simply `return slices.ContainsFunc()`